Implement batch-expiration for objects

krisis opened the pull request:
## Community Contribution License
All community contributions in this pull request are licensed to the project maintainers
under the terms of the [Apache 2 license] (https://www.apache.org/licenses/LICENSE-2.0). 
By creating this pull request I represent that I have the right to license the 
contributions to the project maintainers under the Apache 2 license.

## Description
Based on an initial PR from Harshavardhana -
https://github.com/minio/minio/pull/17792

Companion PRs
- [ ] mc PR: https://github.com/minio/mc/pull/4677
- [x] madmin-go PR: https://github.com/minio/madmin-go/pull/239
- [x] madmin-go PR: https://github.com/minio/madmin-go/pull/248

## Motivation and Context
ILM policy at a bucket-level allows user to expire objects based on prefix or object tags based filtering and object age-based selection criteria. This doesn't work well when the number of days to expire in since object creation is very small, say less than a week.
Batch-expiry feature works on a bucket level. It allows users to express a selection of objects to be expired based on prefix and/or object tags. It differs from ILM in the following ways,
- MinIO's batch jobs subsystem will immediately identify the selected objects and expire them
- Unlike ILM processing, batch-expiry doesn't yield to application I/O
- Runs only once, to completion; If newer objects matching the selection criteria appear in the object namespace, this batch job may be run again.

## How to test this PR?
https://github.com/krisis/batch-expiry-test contains instructions and commands useful to test batch-expiry.

## Types of changes
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Optimization (provides speedup with no functional changes)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
- [ ] Fixes a regression (If yes, please add `commit-id` or `PR #` here)
- [x] Unit tests added/updated
- [ ] Internal documentation updated
- [ ] Create a documentation update request [here](https://github.com/minio/docs/issues/new?label=doc-change,title=Doc+Updated+Needed+For+PR+github.com%2fminio%2fminio%2fpull%2fNNNNN)

Commit:
Implement batch-expiration for objects

Based on an initial PR from Harshavardhana -
https://github.com/minio/minio/pull/17792

Commit:
Make more types available to all batch-jobs

Commit:
Fix Walk's filter function

Commit:
Periodically save in-memory state

Commit:
Implement Purge.RetainVersions

Commit:
Restrict number of filter rules to 10

Commit:
Fixes and stuff

Commit:
Make batch-expire use DeleteObjects API

Commit:
Return batch-expiration metrics on batch-status

Commit:
Export batch metrics via prometheus

Commit:
config: Add support for delay between batch ops

Commit:
batch multiple objects for expiry

Commit:
fix batch expire status tracking

Commit:
fixes for metrics

shtripat commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,634 @@
+	Purge BatchJobExpirePurge `yaml:"purge" json:"purge"`
+}
+
+// Matches returns true if fi matches the filter conditions specified in ef.
```
`if obj matches the filter`?

shtripat commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,634 @@
+			var match bool
+			for t, v := range tagMap {
+				if kv.Match(BatchJobKV{Key: t, Value: v}) {
+					match = true
```
Do we need to break once match found here?

krisis replied:
An object matches the list of tags filter criteria if and only if the object's tags match (glob pattern match) all of them (think AND operation).

shtripat commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,634 @@
+				}
+				// We only need to match x-amz-meta or standardHeaders
+				if kv.Match(BatchJobKV{Key: k, Value: v}) {
+					match = true
```
Same as above for break

krisis replied:
`metadata` filter criteria too can be thought of as an AND operation of `kv.Match` results over all metadata headers specified.

shtripat commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,634 @@
+		}
+	}
+
+	if ef.Size.LesserThan > 0 {
```
I feel for match, by default we can expect object to not match. If conditions met, return true, else false at last. That looks more readable to me.

krisis replied:
This is not possible. For an object is said to have matched the filter criteria only when it matches all of them. Think AND (again). This is also why we can return early if even an object fails to match a single criterion, but we can't declare it as a match unless all the criteria are met.

shtripat commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,634 @@
+	BatchJobExpireDeleted string = "deleted"
+)
+
+// Validate returns nil if f has valid fields, validation error otherwise.
```
`if ef has valid fields` ?

shtripat commented on cmd/batch-job-common-types.go:
```diff
@@ -81,3 +82,37 @@ func (r BatchJobRetry) Validate() error {
+type BatchJobSize int64
+
+// LesserThan returns true if sz is less than s, false otherwise.
+func (s BatchJobSize) LesserThan(sz int64) bool {
```
Semantically I feel `BatchJobSize.LesserThan(arg)` means if current `BatchJobSize` value is lesser than argument value passed. Should we change that way? I think that's the reason I was confused while usage of this function.

krisis replied:
I agree with you that this is confusing. Let me clean this up.

shtripat commented on cmd/batch-job-common-types.go:
```diff
@@ -81,3 +82,37 @@ func (r BatchJobRetry) Validate() error {
+}
+
+// GreaterThan returns true if sz is greater than s, false otherwise.
+func (s BatchJobSize) GreaterThan(sz int64) bool {
```
Same as above

shtripat reviewed (commented):
few minor comments/questions
will verify the changes following the instructions.

shtripat reviewed (approved):
Verified and works as expected.
Don't have any other comments other than prior minor ones.
Verified metrics
```
$ mc admin prometheus metrics myminio bucket | grep _expire_objects 
# HELP minio_bucket_batch_expire_objects Get successfully completed batch job expireobjects
# TYPE minio_bucket_batch_expire_objects gauge
minio_bucket_batch_expire_objects{bucket="bucket",jobId="dQT94EHKcAjERSPS5FikkZ:-1",server="127.0.0.1:9000"} 97
# HELP minio_bucket_batch_expire_objects_failed Get failed batch job expireobjects
# TYPE minio_bucket_batch_expire_objects_failed gauge
minio_bucket_batch_expire_objects_failed{bucket="bucket",jobId="dQT94EHKcAjERSPS5FikkZ:-1",server="127.0.0.1:9000"} 0
```

Commit:
Address review comments

Commit:
Simplify BatchJobSizeFilter related validations

- Added unit tests
- Renamed BatchJobSizeFilter member fields for ease of understanding
  while retaining the exported (yaml) names.

krisis commented on cmd/batch-job-common-types.go:
```diff
@@ -81,3 +83,52 @@ func (r BatchJobRetry) Validate() error {
+// BatchJobSizeFilter supports size based filters - LesserThan and GreaterThan
+type BatchJobSizeFilter struct {
+	UpperBound BatchJobSize `yaml:"lesserThan,omitempty" json:"lesserThan"`
+	LowerBound BatchJobSize `yaml:"greaterThan,omitempty" json:"greaterThan"`
```
@shtripat I have renamed the internal fields so that the methods on them are easy to understand. I have also added unit tests to `BatchJobSizeFilter` methods to cover typical scenarios and to illustrate its usage.

krisis commented on cmd/erasure-server-pool.go:
```diff
@@ -2029,7 +2029,10 @@ func (z *erasureServerPools) Walk(ctx context.Context, bucket, prefix string, re
 
-							versionsSorter(fivs.Versions).reverse()
+							// Note: entry.fileInfoVersions returns versions sorted in reverse chronological order based on ModTime
+							if opts.WalkVersionsSort == WalkVersionsSortAsc {
```
I just realized that this PR introduces support for receiving versions of an object in reverse chronological order of `ModTime`s. To minimize the amount of files that need to be touched, I use the 'zero' value of [`WalkVersionsSortOrder`](https://github.com/minio/minio/pull/17946/files#diff-7410eec4482a2c3834d74dd077538d3b377627a3a50e436a372bdd78934ef837R49) to indicate ascending order as default ordering. I'm drawing attention to this since it affects all consumers of the `ObjectLayer.Walk` method.

klauspost commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,622 @@
+//         - key: content-type
+//           value: image/* # match objects with 'content-type', all values starting with 'image/'
+//       size:
+//         lesserThan: "10MiB" # match objects with size lesser than this value (e.g. 10MiB)
```
Gramatically it is `lessThan`. I can live with it, though.

krisis replied:
I shall address this in this PR. I will need to send a PR to madmin-go for this though.

krisis replied:
https://github.com/minio/madmin-go/pull/248

klauspost commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,622 @@
+		return false
+	}
+
+	if !ef.CreatedBefore.IsZero() && (obj.ModTime.Equal(ef.CreatedBefore) || obj.ModTime.After(ef.CreatedBefore)) {
```
```suggestion
	if !ef.CreatedBefore.IsZero() && !obj.ModTime.Before(ef.CreatedBefore) {
```

klauspost commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,622 @@
+// BatchJobExpireFilter holds all the filters currently supported for batch replication
+type BatchJobExpireFilter struct {
+	OlderThan     time.Duration       `yaml:"olderThan,omitempty" json:"olderThan"`
+	CreatedBefore time.Time           `yaml:"createdBefore,omitempty" json:"createdBefore"`
```
If you want omitempty to work, it must be:
```suggestion
	CreatedBefore *time.Time          `yaml:"createdBefore,omitempty" json:"createdBefore"`
```

klauspost commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,622 @@
+	}
+	if err := ef.Size.Validate(); err != nil {
+		return err
+	}
```
Maybe check if `CreatedBefore` is actually in the past.

klauspost commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,622 @@
+}
+
+func (oiCache objInfoCache) Add(toDel ObjectToDelete, oi *ObjectInfo) {
+	oiCache[fmt.Sprintf("%s-%s", toDel.ObjectName, toDel.VersionID)] = oi
```
So this can become *huge* is there any control over max size?

krisis replied:
This 'cache' would have at most 10 entries in it. We batch up to 10 objects to be expired per batch expiration worker - https://github.com/minio/minio/pull/17946/files#diff-75c25e8d16842df078c17f4d7234993a6982cbae3ecbcb81a5d759f4f398d032R498

klauspost commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,622 @@
+}
+
+func (oiCache objInfoCache) Get(toDel ObjectToDelete) (*ObjectInfo, bool) {
+	oi, ok := oiCache[fmt.Sprintf("%s-%s", toDel.ObjectName, toDel.VersionID)]
+	return oi, ok
```
```suggestion
	return oiCache[fmt.Sprintf("%s-%s", toDel.ObjectName, toDel.VersionID)]
```

klauspost commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,622 @@
+}
+
+// BatchJobExpireV1 v1 of batch expiration job
+type BatchJobExpireV1 struct {
```
Do we realistically have to rewrite for new versions? Maybe just:
```suggestion
type BatchJobExpire struct {
```

klauspost commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,622 @@
+	var i int
+	for toExpire := range expireCh {
+		if i > 0 {
+			if wait := globalBatchConfig.Clone().ExpirationWorkersWait; wait > 0 {
```
Can't you just add a func that returns the value, so the Clone can be avoided?

klauspost commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,622 @@
+	}
+
+	var i int
+	for toExpire := range expireCh {
```
Check if ctx is done?

klauspost commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,622 @@
+}
+
+func batchObjsForDelete(ctx context.Context, r *BatchJobExpireV1, ri *batchJobInfo, job BatchJobRequest, api ObjectLayer, wk *workers.Workers, expireCh <-chan []expireObjInfo) {
+	defer wk.Give()
```
Is the corresponding in the caller? Seems a bit strange.

krisis replied:
### Context (optional)
The batch job starts at the `BatchJobExpire.Start` method. This is where we build a pipeline of objects to distribute among worker goroutines for expiration. At the source of the pipeline, we have `ObjectLayer.Walk` method sending `ObjectInfo`s downstream. In `BatchJobExpire.Start` we apply the various filters specified in the batch expiry yaml on the incoming `ObjectInfo`s and send `[]expireObjInfo` (viz. simply annotated `[]ObjectInfo`) of 'qualified' objects downstream. 
The `batchObjsForDelete` function is that component in this pipeline which spawns worker goroutines and assigns each of them a `[]expireObjInfo`. This is the fan-out point.

### Answer
To answer your question, I use the same `worker.Workers` object which we use for limited concurrency among worker goroutines to manage the goroutine running the `batchObjsForDelete` function too. Yes, the `wk.Take()` is called by its caller right before we call `go batchObjsForDelete(...)`.

klauspost replied:
Maybe move it in here, so it is symmetric - makes the "preconditions" of the function easier to understand.

krisis replied:
It dawned on me that `batchObjsForDelete`'s lifetime is bound to `expireCh`. i.e once `expireCh` channel is closed, we have no more objects to work. Thanks for catching this code smell, which turned out to be superfluous.

klauspost commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,622 @@
+			// DeleteObject(deletePrefix: true) to expire all versions of an object
+			for _, exp := range toExpireAll {
+				var success bool
+				for attempts := 1; attempts <= retryAttempts; attempts++ {
```
Break/return on ctx.Done?

klauspost commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,622 @@
+
+			// DeleteMultiple objects
+			toDelCopy := make([]ObjectToDelete, len(toDel))
+			for attempts := 1; attempts <= retryAttempts; attempts++ {
```
ctx Done

klauspost commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,622 @@
+
+	// Notify expire jobs final status to the configured endpoint
+	buf, _ := json.Marshal(ri)
+	if err := r.Notify(ctx, bytes.NewReader(buf)); err != nil {
```
ctx may be expired here. Maybe
```suggestion
	if err := r.Notify(context.Background(), bytes.NewReader(buf)); err != nil {
```
?

klauspost commented on cmd/batch-handlers.go:
```diff
@@ -431,23 +435,27 @@ func (r *BatchJobReplicateV1) StartFromSource(ctx context.Context, api ObjectLay
 				// persist in-memory state to disk after every 10secs.
 				logger.LogIf(ctx, ri.updateAfter(ctx, api, 10*time.Second, job))
+
+				if wait := globalBatchConfig.Clone().ReplicationWorkersWait; wait > 0 {
```
Avoid clone if feasible.

klauspost commented on cmd/batch-handlers.go:
```diff
@@ -996,16 +1031,20 @@ func (r *BatchJobReplicateV1) Start(ctx context.Context, api ObjectLayer, job Ba
 				// persist in-memory state to disk after every 10secs.
 				logger.LogIf(ctx, ri.updateAfter(ctx, api, 10*time.Second, job))
+
+				if wait := globalBatchConfig.Clone().ReplicationWorkersWait; wait > 0 {
```
same

klauspost reviewed (commented):
Nothing major! Great stuff!

Commit:
Address more review comments

krisis reviewed (commented):
I have addressed most of the review comments. I shall address the remaining in another refresh.

Commit:
Use lessThan instead of lesserThan

typo: grammar

vadmeste commented on internal/config/batch/batch.go:
```diff
@@ -0,0 +1,151 @@
+		return cfg, err
+	}
+
+	if rduration > 0 {
+		cfg.ReplicationWorkersWait = rduration
+	}
+
+	if kduration > 0 {
+		cfg.KeyRotationWorkersWait = kduration
+	}
+
+	if eduration > 0 {
+		cfg.ExpirationWorkersWait = eduration
+	}
```
time.Duration can take a negative value; let's error out if thats the ccase

vadmeste commented on cmd/metrics-v2.go:
```diff
@@ -3247,6 +3249,77 @@ func getClusterHealthMetrics() *MetricsGroup {
+						Subsystem: "batch",
+						Name:      MetricName(jtype + "_objects_failed"),
+						Help:      "Get failed batch job " + jtype + "objects",
+						Type:      gaugeMetric,
```
I think this is a counter metric, not gauge

vadmeste commented on cmd/metrics-v2.go:
```diff
@@ -3247,6 +3249,77 @@ func getClusterHealthMetrics() *MetricsGroup {
+						Subsystem: "batch",
+						Name:      MetricName(jtype + "_objects"),
+						Help:      "Get successfully completed batch job " + jtype + "objects",
+						Type:      gaugeMetric,
```
I think this is a counter metric, not gauge

vadmeste commented on cmd/batch-job-common-types.go:
```diff
@@ -81,3 +83,52 @@ func (r BatchJobRetry) Validate() error {
+
+// BatchJobSizeFilter supports size based filters - LesserThan and GreaterThan
+type BatchJobSizeFilter struct {
+	UpperBound BatchJobSize `yaml:"lessThan,omitempty" json:"lessThan"`
```
I doubt omitempty is necessary  with a struct; I believe only with a pointer

vadmeste commented on cmd/batch-job-common-types.go:
```diff
@@ -81,3 +83,52 @@ func (r BatchJobRetry) Validate() error {
+	if err != nil {
+		return err
+	}
+	*s = BatchJobSize(sz)
```
I guess check if batch job size is >= 0

krisis replied:
https://pkg.go.dev/github.com/dustin/go-humanize#ParseBytes returns an error if input string contains negative values.

vadmeste commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,650 @@
+			return false
+		}
+	default:
+		// we should never come here, Validate should have caught this.
```
maybe let's panic()

harshavardhana replied:
@krisis ^^

krisis replied:
panic() seems a little too much given that we would (safely) ignore rules with invalid filters. I can add an error log to record that something is amiss. Does that sound ok @vadmeste ?

vadmeste commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,650 @@
+			// specified in the filter
+			var match bool
+			for k, v := range obj.UserDefined {
+				if !stringsHasPrefixFold(k, "x-amz-meta-") && !isStandardHeader(k) {
```
Why is this needed ? we can leave the user to specify anything

harshavardhana replied:
@krisis ^^

harshavardhana replied:
WE do not want them to come up internal headers and somehow match them, this is a way to prevent such mistakes.

vadmeste commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,650 @@
+
+	now := time.Now().UTC()
+
+	workerSize, err := strconv.Atoi(env.Get("_MINIO_BATCH_EXPIRATION_WORKERS", strconv.Itoa(runtime.GOMAXPROCS(0)/2)))
```
Don't we have a configuration for this ?

krisis replied:
It was decided that we should only ever increase batch-expiration workers in exceptional cases. Making this available as a configuration will encourage the user behavior we wish to avoid. /cc @harshavardhana

vadmeste commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,650 @@
+		toDel         []expireObjInfo
+	)
+	for result := range results {
+		result := result
```
This doesn't seem needed

vadmeste commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,650 @@
+					stopFn := globalBatchJobsMetrics.trace(batchJobMetricExpire, ri.JobID, attempts)
+					_, err := api.DeleteObject(ctx, exp.Bucket, exp.Name, ObjectOptions{
+						DeletePrefix: true,
+						NoLock:       true,
```
Why no lock ?

krisis replied:
This object is past its expiry and applications aren't expected to access such objects. It is possible though that an application may be accessing this object while it is being expired. This unlikely application failure is acceptable in this context.

klauspost replied:
Probably better with a lock to avoid edge cases.

krisis replied:
Agree. It is better to be safe than sorry here.

vadmeste reviewed (commented):
Few comments, not tested yet

Commit:
Address some more review comments

klauspost commented on cmd/batch-expire.go:
```diff
@@ -0,0 +1,658 @@
+		return err
+	}
+
+	if len(r.Rules) > 10 {
```
Why this (seemingly) arbitrary limit?

krisis replied:
The intention here is to have an upper bound on the number of rules. 10 is definitely arbitrary and small. We could keep it higher, say 1000, similar to the limit on number of rules in an ILM policy.

klauspost replied:
Yeah... 1000 would be insane.

krisis replied:
Should we settle at 100 then? :)

harshavardhana replied:
50 is fine @krisis

krisis replied:
Updated the maximum rules in a batch expiry job to 50.

Commit:
More review comments addressed

harshavardhana reviewed (approved):
LGTM and tested
